File: src/main/java/com/nibado/projects/advent/y2015/Day01.java
package com.nibado.projects.advent.y2015;

import static com.nibado.projects.advent.Util.printAnswer;
import static com.nibado.projects.advent.Util.readResourceAsString;

public class Day01 implements Runnable {

    public static void main(String... argv) {
        new Day01().run();
    }

    @Override
    public void run() {
        String input = readResourceAsString("/2015/day01.txt");

        int floor = 0;
        for (int i = 0; i < input.length(); i++) {
            if (input.charAt(i) == '(') {
                floor++;
            } else if (input.charAt(i) == ')') {
                floor--;
            }
        }

        printAnswer(floor);

        int i;
        floor = 0;
        for (i = 0; i < input.length(); i++) {
            if (input.charAt(i) == '(') {
                floor++;
            } else if (input.charAt(i) == ')') {
                floor--;
            }
            if (floor < 0) {
                break;
            }
        }

        printAnswer(i + 1);
    }
}

File: src/main/kotlin/com/nibado/projects/advent/y2015/Day01.kt
package com.nibado.projects.advent.y2015

import com.nibado.projects.advent.Day
import com.nibado.projects.advent.resourceString

object Day01 : Day {
    private val instructions = resourceString(2015, 1).map { if (it == '(') 1 else -1 }

    override fun part1() = instructions.sum()
    override fun part2(): Int {
        var sum = 0
        var idx = 0
        instructions.mapIndexed { index, i -> index to i }.takeWhile { sum += it.second; idx = it.first;sum >= 0 }

        return idx + 1
    }
}
